pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go

```diff
@@ -734,6 +708,40 @@ func (tp *txnPipeliner) updateLockTrackingInner(
 	}
 }
 
+// maybeCondenseLockFootprint checks whether the lock footprint combined with
+// the in-flight writes exceed the maximum size (as determined by the
+// TrackedWritesMaxSize cluster setting) and attempts to condense the lock
+// footprint if so.
+//
+// ba is optional and only used when logging a warning in case the budget is
+// exceeded.
+func (tp *txnPipeliner) maybeCondenseLockFootprint(ctx context.Context, ba *roachpb.BatchRequest) {
+	maxBytes := TrackedWritesMaxSize.Get(&tp.st.SV)
+	// Compute how many bytes are left for locks after accounting for the
+	// in-flight writes.
+	locksBudget := maxBytes - tp.ifWrites.byteSize()
+	// If we're below budget, there's nothing more to do.
+	if tp.lockFootprint.bytesSize() <= locksBudget {
+		return
+	}
+
+	alreadyCondensed := tp.lockFootprint.condensed
+	condensed := tp.lockFootprint.maybeCondense(ctx, tp.riGen, locksBudget)
+	if condensed && !alreadyCondensed {
+		if tp.condensedIntentsEveryN.ShouldLog() || log.ExpensiveLogEnabled(ctx, 2) {
+			var baDetails string
```

Comment:
redact.StringBuilder, then use `log.Warningf` and include it with %s in the message.